Remove reflection warnings

BrunoBonacci opened the pull request:
It removes the following annoying reflection warnings while using the cider-repl:

```
Reflection warning, compliment/utils.clj:160:15 - reference to field findAll on java.lang.Object can't be resolved.
Reflection warning, compliment/utils.clj:168:51 - reference to field list on java.lang.Object can't be resolved.
Reflection warning, compliment/utils.clj:170:7 - reference to field close on java.lang.Object can't be resolved.
Reflection warning, compliment/utils.clj:168:51 - reference to field list on java.lang.Object can't be resolved.
Reflection warning, compliment/utils.clj:170:7 - reference to field close on java.lang.Object can't be resolved.
```

all test pass with JDK8 and JDK9+

BrunoBonacci commented:
Hi @alexander-yakushev,
when you get chance could you please review and get fix merged?
These reflection warnings are really annoying and cause weird behaviours.

As soon as a new release of `compliment` is available with the fixes, I'll issue a PR to Cider to update the `compliment` release.

alexander-yakushev commented:
Hello Bruno! I understand that these warnings must be annoying, but could you please specify which sort of weird behaviors does it cause? Because basically, you've rewritten the reflection done by the Clojure compiler into a manual one, so nothing should really change in the resulting bytecode.

BrunoBonacci commented:
Hi @alexander-yakushev 

I do have `*warn-on-reflection*` enabled by default on all the projects. When I start the cider-repl I used to get a screenful of warnings which now I have fixed with the following PRs:
  - https://github.com/clojure-emacs/orchard/pull/101
  - https://github.com/clojure-emacs/cider-nrepl/pull/681
  - https://github.com/clojure-emacs/orchard/pull/97 (from hugoduncan)

This PR fixes the last warnings I see.

the strange behaviour I see is that my buffer would jump to an apparently random buffer called "utils.clj" from *another project* when the evaluation triggers the reflection warning. It is very bizarre and I didn't manage to track the issue in cider-nrepl yet.
However, since I removed the reflection from cider-nrepl, orchard and compliment, this issue completely disappeared.

alexander-yakushev commented:
Since you only want to get rid of the warnings here, not remove the reflection itself, it should be enough to put `(set! *warn-on-reflection* false)` at the top of `utils.clj`.  I'd add something like this:

```clj
;; Disable reflection warnings in this file because we must use reflection to
;; support both JDK8 and JDK9+.
(set! *warn-on-reflection* false)
```

Want to do it in this PR or should I commit it on my own?

BrunoBonacci commented:
adding the `(set! *warn-on-reflection* false)` would disable it globally thus disabling it also for the user.
You will need to:
  - capture the previous value
  - `(set! *warn-on-reflection* false)`
  - load/eval the ns
  - restore the initial value

What's your thought on this, do you prefer the current implementation on the one proposed here?

alexander-yakushev commented:
Not really. See https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Compiler.java#L7625. All legitimate ways to load a file (build tools, `require`, etc.) go through this entrypoint which always creates a new frame for the loaded file. So, changing `*warn-on-reflection*` within a file only influences that file.

BrunoBonacci commented:
Great! I didn't know that. Thanks for pointing it out.
I've added the suggested fix and rebased the changes.

alexander-yakushev commented:
Thank you! Pushed this fix in `0.3.11`.
